File: shared/src/commonMain/kotlin/energy/ftw/crypto/Noise.kt
package energy.ftw.crypto

import energy.ftw.concat

const val PROTOCOL_NAME = "Noise_IK_25519_ChaChaPoly_SHA256"
const val DH_BYTES = 32
const val KEY_BYTES = 32
const val HASH_BYTES = 32
const val TAG_BYTES = 16
const val MESSAGE_1_OVERHEAD = DH_BYTES + DH_BYTES + TAG_BYTES + TAG_BYTES
const val MESSAGE_2_OVERHEAD = DH_BYTES + TAG_BYTES

private val MAX_NONCE = (1L shl 62) // stay inside signed Long; 2^64-1 is the spec cap

class NoiseError(message: String, val code: String) : Exception(message)

data class KeyPair(val secretKey: ByteArray, val publicKey: ByteArray) {
    override fun equals(other: Any?): Boolean {
        if (other !is KeyPair) return false
        return secretKey.contentEquals(other.secretKey) && publicKey.contentEquals(other.publicKey)
    }
    override fun hashCode() = secretKey.contentHashCode() * 31 + publicKey.contentHashCode()
}

fun generateKeyPair(): KeyPair {
    val secret = randomBytes(DH_BYTES)
    return KeyPair(secret, X25519.publicKey(secret))
}

fun keyPairFromSecret(secretKey: ByteArray): KeyPair {
    if (secretKey.size != DH_BYTES) {
        throw NoiseError("static key is ${secretKey.size} bytes, need $DH_BYTES", "E_NOISE_KEY")
    }
    return KeyPair(secretKey, X25519.publicKey(secretKey))
}

private fun dh(secretKey: ByteArray, publicKey: ByteArray): ByteArray =
    X25519.sharedSecret(secretKey, publicKey)

private fun hkdfN(chainingKey: ByteArray, ikm: ByteArray, outputs: Int): List<ByteArray> {
    val prk = Hkdf.extract(chainingKey, ikm)
    val okm = Hkdf.expand(prk, ByteArray(0), outputs * HASH_BYTES)
    return (0 until outputs).map { i -> okm.copyOfRange(i * HASH_BYTES, (i + 1) * HASH_BYTES) }
}

private fun nonceBytes(n: Long): ByteArray {
    val out = ByteArray(12)
    var v = n
    for (i in 4 until 12) {
        out[i] = (v and 0xff).toByte()
        v = v ushr 8
    }
    return out
}

class CipherState(key: ByteArray? = null) {
    private var key: ByteArray? = key?.copyOf()
    var nonce: Long = 0
        private set
    private var destroyed = false

    val hasKey: Boolean get() = key != null

    fun setNonce(n: Long) {
        if (n < 0 || n >= MAX_NONCE) {
            throw NoiseError("nonce $n is out of range", "E_NOISE_NONCE_EXHAUSTED")
        }
        nonce = n
    }

    fun encryptWithAd(ad: ByteArray, plaintext: ByteArray): ByteArray {
        assertUsable()
        val k = key ?: return plaintext
        val n = take()
        return ChaChaPoly.encrypt(k, nonceBytes(n), ad, plaintext)
    }

    fun decryptWithAd(ad: ByteArray, ciphertext: ByteArray): ByteArray {
        assertUsable()
        val k = key ?: return ciphertext
        val n = take()
        try {
            return ChaChaPoly.decrypt(k, nonceBytes(n), ad, ciphertext)
        } catch (e: NoiseError) {
            throw if (e.code == "E_NOISE_AUTH") e else NoiseError("authentication failed", "E_NOISE_AUTH")
        }
    }

    fun clone(): CipherState {
        val copy = CipherState(key)
        copy.nonce = nonce
        copy.destroyed = destroyed
        return copy
    }

    fun destroy() {
        key?.fill(0)
        key = null
        destroyed = true
    }

    private fun assertUsable() {
        if (destroyed) throw NoiseError("cipher was destroyed", "E_NOISE_CLOSED")
    }

    private fun take(): Long {
        if (nonce >= MAX_NONCE) throw NoiseError("nonce space exhausted", "E_NOISE_NONCE_EXHAUSTED")
        val n = nonce
        nonce++
        return n
    }
}

private class SymmetricState {
    var ck: ByteArray
    var h: ByteArray
    var cipher = CipherState()

    init {
        val name = PROTOCOL_NAME.encodeToByteArray()
        h = if (name.size <= HASH_BYTES) {
            concat(name, ByteArray(HASH_BYTES - name.size))
        } else {
            Sha256.hash(name)
        }
        ck = h.copyOf()
    }

    fun mixHash(data: ByteArray) {
        h = Sha256.hash(concat(h, data))
    }

    fun mixKey(ikm: ByteArray) {
        val parts = hkdfN(ck, ikm, 2)
        ck = parts[0]
        cipher = CipherState(parts[1])
    }

    fun encryptAndHash(plaintext: ByteArray): ByteArray {
        val ciphertext = cipher.encryptWithAd(h, plaintext)
        mixHash(ciphertext)
        return ciphertext
    }

    fun decryptAndHash(ciphertext: ByteArray): ByteArray {
        val plaintext = cipher.decryptWithAd(h, ciphertext)
        mixHash(ciphertext)
        return plaintext
    }

    fun clone(): SymmetricState {
        val copy = SymmetricState()
        copy.ck = ck.copyOf()
        copy.h = h.copyOf()
        copy.cipher = cipher.clone()
        return copy
    }

    fun split(): Pair<CipherState, CipherState> {
        val parts = hkdfN(ck, ByteArray(0), 2)
        return CipherState(parts[0]) to CipherState(parts[1])
    }
}

class HandshakeResult(
    val send: CipherState,
    val recv: CipherState,
    val handshakeHash: ByteArray,
    val remoteStatic: ByteArray,
)

private enum class Step { Write1, Read1, Write2, Read2, Done, Split }

class HandshakeState private constructor(
    private val initiator: Boolean,
    staticKey: KeyPair,
    remoteStatic: ByteArray?,
    prologue: ByteArray,
    private val fixedEphemeral: KeyPair?,
) {
    private var sym = SymmetricState()
    private val s = staticKey
    private var e: KeyPair? = null
    private var rs: ByteArray? = remoteStatic
    private var re: ByteArray? = null
    private var step: Step = if (initiator) Step.Write1 else Step.Read1

    init {
        sym.mixHash(prologue)
        if (initiator) {
            val remote = remoteStatic
            if (remote == null || remote.size != DH_BYTES) {
                throw NoiseError("IK needs the responder static key up front", "E_NOISE_KEY")
            }
            rs = remote
            sym.mixHash(remote)
        } else {
            sym.mixHash(s.publicKey)
        }
    }

    val isComplete: Boolean get() = step == Step.Done

    fun writeMessage(payload: ByteArray = ByteArray(0)): ByteArray {
        if (initiator && step == Step.Write1) return writeMessage1(payload)
        if (!initiator && step == Step.Write2) return writeMessage2(payload)
        throw NoiseError("cannot write at step $step", "E_NOISE_STATE")
    }

    fun readMessage(message: ByteArray): ByteArray {
        if (!initiator && step == Step.Read1) return readMessage1(message)
        if (initiator && step == Step.Read2) return readMessage2(message)
        throw NoiseError("cannot read at step $step", "E_NOISE_STATE")
    }

    fun split(): HandshakeResult {
        if (step == Step.Split) throw NoiseError("handshake already split", "E_NOISE_STATE")
        if (step != Step.Done) throw NoiseError("handshake is at step $step", "E_NOISE_STATE")
        val (c1, c2) = sym.split()
        val result = HandshakeResult(
            send = if (initiator) c1 else c2,
            recv = if (initiator) c2 else c1,
            handshakeHash = sym.h,
            remoteStatic = rs!!,
        )
        step = Step.Split
        return result
    }

    private fun writeMessage1(payload: ByteArray): ByteArray {
        e = fixedEphemeral ?: generateKeyPair()
        val eph = e!!
        sym.mixHash(eph.publicKey)
        sym.mixKey(dh(eph.secretKey, rs!!))
        val encStatic = sym.encryptAndHash(s.publicKey)
        sym.mixKey(dh(s.secretKey, rs!!))
        val encPayload = sym.encryptAndHash(payload)
        step = Step.Read2
        return concat(eph.publicKey, encStatic, encPayload)
    }

    private fun readMessage1(message: ByteArray): ByteArray {
        val encStaticEnd = DH_BYTES + DH_BYTES + TAG_BYTES
        if (message.size < encStaticEnd + TAG_BYTES) {
            throw NoiseError("handshake message 1 is ${message.size} bytes", "E_NOISE_MESSAGE")
        }
        val trial = sym.clone()
        val reIn = message.copyOfRange(0, DH_BYTES)
        trial.mixHash(reIn)
        trial.mixKey(dh(s.secretKey, reIn))
        val rsIn = trial.decryptAndHash(message.copyOfRange(DH_BYTES, encStaticEnd))
        trial.mixKey(dh(s.secretKey, rsIn))
        val payload = trial.decryptAndHash(message.copyOfRange(encStaticEnd, message.size))
        sym = trial
        re = reIn
        rs = rsIn
        step = Step.Write2
        return payload
    }

    private fun writeMessage2(payload: ByteArray): ByteArray {
        e = fixedEphemeral ?: generateKeyPair()
        val eph = e!!
        sym.mixHash(eph.publicKey)
        sym.mixKey(dh(eph.secretKey, re!!))
        sym.mixKey(dh(eph.secretKey, rs!!))
        val encPayload = sym.encryptAndHash(payload)
        step = Step.Done
        return concat(eph.publicKey, encPayload)
    }

    private fun readMessage2(message: ByteArray): ByteArray {
        if (message.size < DH_BYTES + TAG_BYTES) {
            throw NoiseError("handshake message 2 is ${message.size} bytes", "E_NOISE_MESSAGE")
        }
        val trial = sym.clone()
        val reIn = message.copyOfRange(0, DH_BYTES)
        trial.mixHash(reIn)
        trial.mixKey(dh(e!!.secretKey, reIn))
        trial.mixKey(dh(s.secretKey, reIn))
        val payload = trial.decryptAndHash(message.copyOfRange(DH_BYTES, message.size))
        sym = trial
        re = reIn
        step = Step.Done
        return payload
    }

    companion object {
        fun initiator(
            staticKey: KeyPair,
            remoteStatic: ByteArray,
            prologue: ByteArray = ByteArray(0),
            ephemeral: KeyPair? = null,
        ): HandshakeState = HandshakeState(true, staticKey, remoteStatic, prologue, ephemeral)

        fun responder(
            staticKey: KeyPair,
            prologue: ByteArray = ByteArray(0),
            ephemeral: KeyPair? = null,
        ): HandshakeState = HandshakeState(false, staticKey, null, prologue, ephemeral)
    }
}

File: appleApp/Sources/FTWKit/Crypto/Noise.swift
import Foundation

/// Noise_IK_25519_ChaChaPoly_SHA256.
///
/// IK because the initiator already knows the responder's static key: the
/// app reads it optically off the box's QR code, so the trust anchor never
/// travels through Sourceful's cloud. A hostile relay can drop frames but
/// cannot present itself as a box.
///
///     IK:
///       <- s
///       ...
///       -> e, es, s, ss
///       <- e, ee, se
///
/// Byte-identical to the TypeScript client and the Go box, and checked
/// against the Cacophony vectors and the shared interop vectors in the tests.
public enum Noise {
    public static let protocolName = "Noise_IK_25519_ChaChaPoly_SHA256"
    public static let dhBytes = 32
    public static let hashBytes = 32
    public static let tagBytes = 16
    /// 2^64 - 1 is reserved by the spec and must never encrypt.
    public static let maxNonce = UInt64.max

    public static let message1Overhead = dhBytes + dhBytes + tagBytes + tagBytes
    public static let message2Overhead = dhBytes + tagBytes

    /// Failures here are local: they never cross the wire. The carrier maps
    /// them to prose; this layer only says which thing went wrong.
    public struct NoiseError: Error, Equatable {
        public let code: String
        public let message: String
    }

    /// Noise HKDF: one extract over the chaining key, then expand with empty
    /// info. Splitting one expand is byte-identical to the chained HMACs.
    static func hkdf2(_ chainingKey: Bytes, _ ikm: Bytes) -> (Bytes, Bytes) {
        let prk = Primitives.hmacSHA256(key: chainingKey, ikm)
        let t1 = Primitives.hmacSHA256(key: prk, [1])
        let t2 = Primitives.hmacSHA256(key: prk, t1 + [2])
        return (t1, t2)
    }

    static func dh(_ secret: Bytes, _ peer: Bytes) throws -> Bytes {
        do {
            return try Primitives.x25519(secret: secret, peer: peer)
        } catch {
            throw NoiseError(code: "E_NOISE_DH", message: "X25519 failed")
        }
    }

    /// ChaChaPoly's 96-bit nonce, Noise-encoded: four zero bytes, then the
    /// 64-bit counter little-endian.
    static func nonceBytes(_ n: UInt64) -> Bytes {
        var out = Bytes(repeating: 0, count: 12)
        for i in 0..<8 { out[4 + i] = UInt8(truncatingIfNeeded: n >> (8 * UInt64(i))) }
        return out
    }
}

/// One direction's key and counter.
///
/// The counter never wraps. Reusing a (key, nonce) pair under
/// ChaCha20-Poly1305 hands an observer the XOR of two plaintexts and makes the
/// authenticator forgeable, so exhaustion throws and the session dies.
public final class CipherState {
    private var key: Bytes?
    public private(set) var nonce: UInt64 = 0
    private var destroyed = false

    public init(key: Bytes? = nil) {
        self.key = key
    }

    public var hasKey: Bool { key != nil }

    /// Jump the counter, for a carrier that may reorder or drop. Replay
    /// protection lives in the transport, which decides what is acceptable.
    public func setNonce(_ n: UInt64) throws {
        guard n < Noise.maxNonce else {
            throw Noise.NoiseError(code: "E_NOISE_NONCE_EXHAUSTED", message: "nonce \(n) is out of range")
        }
        nonce = n
    }

    public func encrypt(ad: Bytes, _ plaintext: Bytes) throws -> Bytes {
        try assertUsable()
        // Before the first mixKey there is no key and the spec passes data
        // through. A destroyed cipher is a separate flag so it can never
        // quietly emit plaintext.
        guard let key else { return plaintext }
        let n = try take()
        return try Primitives.chachaSeal(key: key, nonce: Noise.nonceBytes(n), ad: ad, plaintext: plaintext)
    }

    public func decrypt(ad: Bytes, _ ciphertext: Bytes) throws -> Bytes {
        try assertUsable()
        guard let key else { return ciphertext }
        let n = try take()
        do {
            return try Primitives.chachaOpen(key: key, nonce: Noise.nonceBytes(n), ad: ad, ciphertext: ciphertext)
        } catch {
            // Never say more than this. Which check failed is exactly what a
            // chosen-ciphertext attacker is probing for.
            throw Noise.NoiseError(code: "E_NOISE_AUTH", message: "authentication failed")
        }
    }

    func copy() -> CipherState {
        let c = CipherState(key: key)
        c.nonce = nonce
        c.destroyed = destroyed
        return c
    }

    /// Wipe the key. A closed session must not leave one reachable.
    public func destroy() {
        if var k = key { wipe(&k) }
        key = nil
        destroyed = true
    }

    private func assertUsable() throws {
        if destroyed { throw Noise.NoiseError(code: "E_NOISE_CLOSED", message: "cipher was destroyed") }
    }

    private func take() throws -> UInt64 {
        guard nonce < Noise.maxNonce else {
            throw Noise.NoiseError(code: "E_NOISE_NONCE_EXHAUSTED", message: "nonce space exhausted")
        }
        defer { nonce += 1 }
        return nonce
    }
}

/// Chaining key, handshake hash, and the cipher over handshake payloads.
final class SymmetricState {
    var ck: Bytes
    var h: Bytes
    var cipher = CipherState()

    init() {
        let name = Array(Noise.protocolName.utf8)
        // The name is exactly 32 bytes, so it is used verbatim; a longer one
        // would be hashed.
        h = name.count <= Noise.hashBytes ? name + Bytes(repeating: 0, count: Noise.hashBytes - name.count) : Primitives.sha256(name)
        ck = h
    }

    func mixHash(_ data: Bytes) {
        h = Primitives.sha256(h + data)
    }

    func mixKey(_ ikm: Bytes) {
        let (next, temp) = Noise.hkdf2(ck, ikm)
        ck = next
        cipher = CipherState(key: temp)
    }

    func encryptAndHash(_ plaintext: Bytes) throws -> Bytes {
        let ct = try cipher.encrypt(ad: h, plaintext)
        mixHash(ct)
        return ct
    }

    func decryptAndHash(_ ciphertext: Bytes) throws -> Bytes {
        let pt = try cipher.decrypt(ad: h, ciphertext)
        mixHash(ciphertext)
        return pt
    }

    func copy() -> SymmetricState {
        let c = SymmetricState()
        c.ck = ck
        c.h = h
        c.cipher = cipher.copy()
        return c
    }

    func split() -> (CipherState, CipherState) {
        let (k1, k2) = Noise.hkdf2(ck, [])
        return (CipherState(key: k1), CipherState(key: k2))
    }
}

public struct HandshakeResult {
    /// Encrypts what we send.
    public let send: CipherState
    /// Decrypts what we receive.
    public let recv: CipherState
    public let handshakeHash: Bytes
    /// The peer's static key, now authenticated rather than merely claimed.
    public let remoteStatic: Bytes
}

/// The IK state machine. Anything out of order throws rather than producing
/// a message the peer cannot read.
public final class HandshakeState {
    private enum Step { case write1, read1, write2, read2, done, split }

    private var sym = SymmetricState()
    private let s: Primitives.KeyPair
    private var e: Primitives.KeyPair?
    private var rs: Bytes?
    private var re: Bytes?
    private let initiator: Bool
    private let fixedEphemeral: Primitives.KeyPair?
    private var step: Step

    /// `ephemeral` is for test vectors only: generating it is the whole
    /// source of forward secrecy.
    public static func initiator(staticKey: Primitives.KeyPair, remoteStatic: Bytes, prologue: Bytes = [], ephemeral: Primitives.KeyPair? = nil) throws -> HandshakeState {
        guard remoteStatic.count == Noise.dhBytes else {
            throw Noise.NoiseError(code: "E_NOISE_KEY", message: "IK needs the responder static key up front")
        }
        return HandshakeState(initiator: true, staticKey: staticKey, remoteStatic: remoteStatic, prologue: prologue, ephemeral: ephemeral)
    }

    public static func responder(staticKey: Primitives.KeyPair, prologue: Bytes = [], ephemeral: Primitives.KeyPair? = nil) -> HandshakeState {
        HandshakeState(initiator: false, staticKey: staticKey, remoteStatic: nil, prologue: prologue, ephemeral: ephemeral)
    }

    private init(initiator: Bool, staticKey: Primitives.KeyPair, remoteStatic: Bytes?, prologue: Bytes, ephemeral: Primitives.KeyPair?) {
        self.initiator = initiator
        s = staticKey
        fixedEphemeral = ephemeral
        step = initiator ? .write1 : .read1
        sym.mixHash(prologue)
        // IK's pre-message `<- s`: both ends hash the responder's static key
        // before a byte moves, so a relay substituting its own key fails at
        // the first decryption.
        if initiator, let remoteStatic {
            rs = remoteStatic
            sym.mixHash(remoteStatic)
        } else {
            sym.mixHash(staticKey.publicKey)
        }
    }

    public var isComplete: Bool { step == .done }

    public func writeMessage(_ payload: Bytes = []) throws -> Bytes {
        if initiator && step == .write1 { return try writeMessage1(payload) }
        if !initiator && step == .write2 { return try writeMessage2(payload) }
        throw Noise.NoiseError(code: "E_NOISE_STATE", message: "cannot write at this step")
    }

    public func readMessage(_ message: Bytes) throws -> Bytes {
        if !initiator && step == .read1 { return try readMessage1(message) }
        if initiator && step == .read2 { return try readMessage2(message) }
        throw Noise.NoiseError(code: "E_NOISE_STATE", message: "cannot read at this step")
    }

    /// Ends the handshake and hands over the keys, once. Splitting twice
    /// would mint two send ciphers with the same key at nonce 0.
    public func split() throws -> HandshakeResult {
        guard step == .done else {
            throw Noise.NoiseError(code: "E_NOISE_STATE", message: step == .split ? "handshake already split" : "handshake is not done")
        }
        let (c1, c2) = sym.split()
        step = .split
        return HandshakeResult(
            send: initiator ? c1 : c2,
            recv: initiator ? c2 : c1,
            handshakeHash: sym.h,
            remoteStatic: rs ?? []
        )
    }

    // -> e, es, s, ss
    private func writeMessage1(_ payload: Bytes) throws -> Bytes {
        let e = fixedEphemeral ?? Primitives.generateKeyPair()
        self.e = e
        let rs = self.rs!
        sym.mixHash(e.publicKey)
        sym.mixKey(try Noise.dh(e.secretKey, rs))
        let encStatic = try sym.encryptAndHash(s.publicKey)
        sym.mixKey(try Noise.dh(s.secretKey, rs))
        let encPayload = try sym.encryptAndHash(payload)
        step = .read2
        return e.publicKey + encStatic + encPayload
    }

    private func readMessage1(_ message: Bytes) throws -> Bytes {
        let encStaticEnd = Noise.dhBytes + Noise.dhBytes + Noise.tagBytes
        guard message.count >= encStaticEnd + Noise.tagBytes else {
            throw Noise.NoiseError(code: "E_NOISE_MESSAGE", message: "handshake message 1 is \(message.count) bytes")
        }
        // Commit on success: a message that fails to authenticate must leave
        // no trace, or one stray frame ends every handshake still waiting.
        let sym = self.sym.copy()
        let re = Array(message[0..<Noise.dhBytes])
        sym.mixHash(re)
        sym.mixKey(try Noise.dh(s.secretKey, re))
        let rs = try sym.decryptAndHash(Array(message[Noise.dhBytes..<encStaticEnd]))
        sym.mixKey(try Noise.dh(s.secretKey, rs))
        let payload = try sym.decryptAndHash(Array(message[encStaticEnd...]))
        self.sym = sym
        self.re = re
        self.rs = rs
        step = .write2
        return payload
    }

    // <- e, ee, se
    private func writeMessage2(_ payload: Bytes) throws -> Bytes {
        let e = fixedEphemeral ?? Primitives.generateKeyPair()
        self.e = e
        sym.mixHash(e.publicKey)
        sym.mixKey(try Noise.dh(e.secretKey, re!))
        sym.mixKey(try Noise.dh(e.secretKey, rs!))
        let encPayload = try sym.encryptAndHash(payload)
        step = .done
        return e.publicKey + encPayload
    }

    private func readMessage2(_ message: Bytes) throws -> Bytes {
        guard message.count >= Noise.dhBytes + Noise.tagBytes else {
            throw Noise.NoiseError(code: "E_NOISE_MESSAGE", message: "handshake message 2 is \(message.count) bytes")
        }
        // Mixing runs against a copy that becomes the state only once the tag
        // verifies. Two phones connecting at once read each other's 48-byte
        // replies on a shared relay room; mixing a stray one would poison the
        // state and the genuine reply could never authenticate.
        let sym = self.sym.copy()
        let re = Array(message[0..<Noise.dhBytes])
        sym.mixHash(re)
        sym.mixKey(try Noise.dh(e!.secretKey, re))
        sym.mixKey(try Noise.dh(s.secretKey, re))
        let payload = try sym.decryptAndHash(Array(message[Noise.dhBytes...]))
        self.sym = sym
        self.re = re
        step = .done
        return payload
    }
}
